codex/stores/maintenance.nim

```diff
@@ -10,92 +10,323 @@
+  except CancelledError as err:
+    raise err
+  except CatchableError as err:
+    error "Unexpected error during dataset deletion", msg = err.msg, treeCid = treeCid
+
+proc listDatasetMetadata*(
+  self: DatasetMaintainer
+): Future[?!AsyncIter[(Cid, DatasetMetadata)]] {.async.} =
+  without queryKey =? createDatasetMetadataQueryKey(), err:
+    return failure(err)
+
+  without queryIter =? await query[DatasetMetadata](self.metaDs, Query.init(queryKey)), err:
+    error "Unable to execute block expirations query", err = err.msg
+    return failure(err)
+
+  without asyncQueryIter =? await queryIter.toAsyncIter(), err:
+    error "Unable to convert QueryIter to AsyncIter", err = err.msg
+    return failure(err)
+
+  let
+    filteredIter = await asyncQueryIter.filterSuccess()
+
+    datasetMdIter = await mapFilter[KeyVal[DatasetMetadata], (Cid, DatasetMetadata)](filteredIter,
+      proc (kv: KeyVal[DatasetMetadata]): Future[?(Cid, DatasetMetadata)] {.async.} =
+        without cid =? Cid.init(kv.key.value).mapFailure, err:
+          error "Failed decoding cid", err = err.msg
+          return (Cid, DatasetMetadata).none
+
+        (cid, kv.value).some
+    )
+
+  success(datasetMdIter)
+
+
+proc checkDatasets(self: DatasetMaintainer): Future[?!void] {.async.} =
+  without iter =? await self.listDatasetMetadata(), err:
+    return failure(err)
 
-  var numberReceived = 0
-  for beFut in iter:
-    let be = await beFut
-    inc numberReceived
-    await self.processBlockExpiration(be)
-    await sleepAsync(1.millis) # cooperative scheduling
+  for fut in iter:
+    let (treeCid, datasetMd) = await fut
 
-  # If we received fewer blockExpirations from the iterator than we asked for,
-  # We're at the end of the dataset and should start from 0 next time.
-  if numberReceived < self.numberOfBlocksPerInterval:
-    self.offset = 0
+    if (datasetMd.expiry < self.clock.now) and
+        (datasetMd.checkpoint.timestamp + self.retryDelay.seconds < self.clock.now):
+      asyncSpawn self.superviseDatasetDeletion(treeCid, datasetMd).track(self)
+    else:
+      trace "Item either not expired or expired but already in maintenance", treeCid, expiry = datasetMd.expiry, timestamp = datasetMd.checkpoint.timestamp
+  success()
 
-proc start*(self: BlockMaintainer) =
+proc start*(self: DatasetMaintainer) =
   proc onTimer(): Future[void] {.async.} =
     try:
-      await self.runBlockCheck()
-    except CancelledError as error:
-      raise error
-    except CatchableError as exc:
-      error "Unexpected exception in BlockMaintainer.onTimer(): ", msg=exc.msg
+      if err =? (await self.checkDatasets()).errorOption:
+        error "Error when checking datasets", msg = err.msg
+    except CancelledError as err:
+      raise err
+    except CatchableError as err:
+      error "Error when checking datasets", msg = err.msg
 
-  self.timer.start(onTimer, self.interval)
+  if self.interval.seconds > 0:
+    self.timer.start(onTimer, self.interval)
```

Comment:
I would avoid using the `timer` altogether and just use a regular async `for` here?

Reply:
What you mean is this?

```nim
while (self.notStopped):
   doStuff()
   await asyncSleep(self.delay)
```

Reply:
For now I left the timer since I didn't get the confirmation if what I proposed above is what you mean.